File: backend/app/api/routes/assignments.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import require_admin
from app.core.database import get_db
from app.crud import assignment as assignment_crud
from app.crud import audit as audit_crud
from app.crud import submission as submission_crud
from app.crud import user as user_crud
from app.models.enums import UserRole
from app.models.user import User
from app.schemas.assignment import (
    AssignmentOut,
    ConflictExclusionCreate,
    ConflictExclusionOut,
    GenerateAssignmentsRequest,
    GenerateAssignmentsResult,
    ManualAssignmentCreate,
)
from app.services import assignment_service

router = APIRouter(prefix="/assignments", tags=["assignments"])


def _to_out(assignment) -> AssignmentOut:
    return AssignmentOut(
        id=assignment.id,
        submission=assignment.submission,
        judge=assignment.judge,
        source=assignment.source,
        assigned_at=assignment.assigned_at,
        evaluation_status=assignment.evaluation.status.value if assignment.evaluation else "not_started",
    )


@router.get("", response_model=list[AssignmentOut])
def list_assignments(
    judge_id: int | None = None,
    submission_id: int | None = None,
    db: Session = Depends(get_db),
    _: User = Depends(require_admin),
) -> list[AssignmentOut]:
    return [_to_out(a) for a in assignment_crud.list_all(db, judge_id, submission_id)]


@router.post("/generate", response_model=GenerateAssignmentsResult)
def generate(
    payload: GenerateAssignmentsRequest, db: Session = Depends(get_db), admin: User = Depends(require_admin)
) -> GenerateAssignmentsResult:
    result = assignment_service.generate_assignments(db, payload, admin.id)
    audit_crud.create(db, admin.id, "generate", "assignment", None, result.model_dump())
    return result


@router.post("", response_model=AssignmentOut, status_code=status.HTTP_201_CREATED)
def create_manual_assignment(
    payload: ManualAssignmentCreate, db: Session = Depends(get_db), admin: User = Depends(require_admin)
) -> AssignmentOut:
    if not submission_crud.get(db, payload.submission_id):
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Submission not found")
    judge = user_crud.get_by_id(db, payload.judge_id)
    if not judge or judge.role != UserRole.JUDGE:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Judge not found")
    if assignment_crud.exists(db, payload.submission_id, payload.judge_id):
        raise HTTPException(status.HTTP_409_CONFLICT, detail="This judge is already assigned to this submission")

    assignment = assignment_crud.create(db, payload.submission_id, payload.judge_id, admin.id, source="manual")
    audit_crud.create(
        db, admin.id, "create", "assignment", assignment.id,
        {"submission_id": payload.submission_id, "judge_id": payload.judge_id},
    )
    return _to_out(assignment)


@router.delete("/{assignment_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_assignment(assignment_id: int, db: Session = Depends(get_db), admin: User = Depends(require_admin)) -> None:
    assignment = assignment_crud.get(db, assignment_id)
    if not assignment:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Assignment not found")
    audit_crud.create(
        db, admin.id, "delete", "assignment", assignment.id,
        {"submission_id": assignment.submission_id, "judge_id": assignment.judge_id},
    )
    assignment_crud.delete(db, assignment)


@router.get("/conflicts", response_model=list[ConflictExclusionOut])
def list_conflicts(db: Session = Depends(get_db), _: User = Depends(require_admin)) -> list[ConflictExclusionOut]:
    return [ConflictExclusionOut.model_validate(c) for c in assignment_crud.list_conflicts(db)]


@router.post("/conflicts", response_model=ConflictExclusionOut, status_code=status.HTTP_201_CREATED)
def create_conflict(
    payload: ConflictExclusionCreate, db: Session = Depends(get_db), admin: User = Depends(require_admin)
) -> ConflictExclusionOut:
    conflict = assignment_crud.create_conflict(db, payload.submission_id, payload.judge_id, payload.reason)
    audit_crud.create(db, admin.id, "create", "conflict_exclusion", conflict.id, {})
    return ConflictExclusionOut.model_validate(conflict)


@router.delete("/conflicts/{conflict_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_conflict(conflict_id: int, db: Session = Depends(get_db), admin: User = Depends(require_admin)) -> None:
    conflict = assignment_crud.get_conflict(db, conflict_id)
    if not conflict:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail="Conflict exclusion not found")
    audit_crud.create(db, admin.id, "delete", "conflict_exclusion", conflict.id, {})
    assignment_crud.delete_conflict(db, conflict)

File: workers/src/routes/assignments.ts
import { Hono } from "hono";

import { conflict, notFound } from "../http";
import { intPathParam, intQueryParam, parseOrThrow, readJson } from "../lib/validate";
import { requireAdmin, requireUser, type AppEnv } from "../middleware/auth";
import * as assignmentRepo from "../repo/assignment";
import * as auditRepo from "../repo/audit";
import * as submissionRepo from "../repo/submission";
import * as userRepo from "../repo/user";
import {
  conflictExclusionSchema,
  generateAssignmentsSchema,
  manualAssignmentSchema,
} from "../schemas/assignment";
import { generateAssignments } from "../services/assignments";

/** Matches AssignmentOut in backend/app/schemas/assignment.py. */
function assignmentOut(row: assignmentRepo.AssignmentJoined) {
  return {
    id: row.id,
    submission: {
      id: row.submission_id,
      project_title: row.submission_title,
      team_identifier: row.submission_team,
    },
    judge: { id: row.judge_id, email: row.judge_email, full_name: row.judge_full_name },
    source: row.source,
    assigned_at: row.assigned_at,
    evaluation_status: row.evaluation_status ?? "not_started",
  };
}

export const assignmentRoutes = new Hono<AppEnv>();

assignmentRoutes.use("*", requireUser, requireAdmin);

assignmentRoutes.get("/conflicts", async (c) =>
  c.json(await assignmentRepo.listConflicts(c.env.DB))
);

assignmentRoutes.post("/conflicts", async (c) => {
  const payload = parseOrThrow(conflictExclusionSchema, await readJson(c.req), "body");
  const row = await assignmentRepo.createConflict(
    c.env.DB,
    payload.submission_id,
    payload.judge_id,
    payload.reason ?? null
  );
  const admin = c.get("user");
  await auditRepo.create(c.env.DB, admin.id, "create", "conflict_exclusion", row.id, {});
  return c.json(row, 201);
});

assignmentRoutes.delete("/conflicts/:conflict_id", async (c) => {
  const conflictId = intPathParam("conflict_id", c.req.param("conflict_id"));
  const row = await assignmentRepo.getConflict(c.env.DB, conflictId);
  if (!row) throw notFound("Conflict exclusion not found");

  const admin = c.get("user");
  await auditRepo.create(c.env.DB, admin.id, "delete", "conflict_exclusion", conflictId, {});
  await assignmentRepo.deleteConflict(c.env.DB, conflictId);
  return c.body(null, 204);
});

assignmentRoutes.post("/generate", async (c) => {
  const payload = parseOrThrow(generateAssignmentsSchema, await readJson(c.req), "body");
  const admin = c.get("user");
  const result = await generateAssignments(c.env.DB, payload, admin.id);
  await auditRepo.create(c.env.DB, admin.id, "generate", "assignment", null, { ...result });
  return c.json(result);
});

assignmentRoutes.get("/", async (c) => {
  const judgeId = intQueryParam("judge_id", c.req.query("judge_id"));
  const submissionId = intQueryParam("submission_id", c.req.query("submission_id"));
  const rows = await assignmentRepo.listAll(c.env.DB, judgeId, submissionId);
  return c.json(rows.map(assignmentOut));
});

assignmentRoutes.post("/", async (c) => {
  const payload = parseOrThrow(manualAssignmentSchema, await readJson(c.req), "body");

  if (!(await submissionRepo.get(c.env.DB, payload.submission_id))) {
    throw notFound("Submission not found");
  }
  const judge = await userRepo.getById(c.env.DB, payload.judge_id);
  if (!judge || judge.role !== "judge") throw notFound("Judge not found");
  if (await assignmentRepo.exists(c.env.DB, payload.submission_id, payload.judge_id)) {
    throw conflict("This judge is already assigned to this submission");
  }

  const admin = c.get("user");
  const created = await assignmentRepo.create(
    c.env.DB,
    payload.submission_id,
    payload.judge_id,
    admin.id,
    "manual"
  );
  if (!created) throw notFound("Assignment not found");

  await auditRepo.create(c.env.DB, admin.id, "create", "assignment", created.id, {
    submission_id: payload.submission_id,
    judge_id: payload.judge_id,
  });
  return c.json(assignmentOut(created), 201);
});

assignmentRoutes.delete("/:assignment_id", async (c) => {
  const assignmentId = intPathParam("assignment_id", c.req.param("assignment_id"));
  const row = await assignmentRepo.get(c.env.DB, assignmentId);
  if (!row) throw notFound("Assignment not found");

  const admin = c.get("user");
  await auditRepo.create(c.env.DB, admin.id, "delete", "assignment", assignmentId, {
    submission_id: row.submission_id,
    judge_id: row.judge_id,
  });
  await assignmentRepo.remove(c.env.DB, assignmentId);
  return c.body(null, 204);
});
